espwrap/adaptors/sendgrid_v3.py

```diff
@@ -0,0 +1,201 @@
+            if self.send_at:
+                message.send_at = SendAt(self.send_at)
+
+            #Category
+            if self.tags:
+                message.Category = Category(self.tags)
+
+            #IP Pool
+            if self.ip_pool:
+                message.ip_pool_name = IpPoolName(self.ip_pool)
+
+            #CC
+            if self.cc_list:
+                list_cc = []
+                for email in self.cc_list:
+                    list_cc.append(Cc(email))
+                message.cc = list_cc
+
+            #BCC
+            if self.bcc_list:
+                list_bcc = []
+                for email in self.bcc_list:
+                    list_bcc.append(Bcc(email))
+                message.bcc = list_bcc
+        
+            #Global Subs
+            for key, val in self.global_merge_vars.items():
+                new_key = '{1}{0}{2}'.format(key, *self.delimiters)
+                message.substitution = Substitution(new_key, val)
+
+            #Content (Text)
+            if self.body[MIMETYPE_TEXT]:
+                content = Content(
+                    MimeType.text,
+                    self.body[MIMETYPE_TEXT]
+                )
+                message.content = content
+
+            #Content (Html)
+            if self.body[MIMETYPE_HTML]:
+                content = Content(
+                    MimeType.html,
+                    self.body[MIMETYPE_HTML]
+                )
+                message.content = content
+                
+            #Attachment
+            if self.attachments:
+                for file_name, file_path_or_string in self.attachments.iteritems():
+                    attachment = Attachment()
+                    encoded = base64.b64encode(str(file_path_or_string))
+                    attachment.file_content = FileContent(encoded)
+                    attachment.file_name = FileName(file_name)
+                message.attachment = attachment
+
+            #Webhook Data
+            if self.webhook_data:
+                for key, val in self.webhook_data.items():
+                    message.custom_arg = CustomArg(key, val)
+            if self.template_name:
+                message.custom_arg = CustomArg('template_name', self.template_name)
+                    
+            #Tracking
+            tracking_settings = TrackingSettings()            
+            #Opens
+            if self.track_opens:
+                tracking_settings.open_tracking = OpenTracking(
+                    self.track_opens,
+                    OpenTrackingSubstitutionTag("open_tracking")
+                )
+            #Clicks
+            if self.track_clicks:
+                tracking_settings.click_tracking = ClickTracking(
+                    self.track_clicks,
+                    self.track_clicks
+                )
+            message.tracking_settings = tracking_settings
+            
+            #Personalizations (DO THIS LAST)
+            for subgrp in to_send:
+                p=0
```

Comment:
you can have python do this automatically with `enumerate`
eg: `for p, recip in enumerate(subgrp)` -- I would use a better variable name than p, which technically isn't PEP8 compliant.

Reply:
nice. i set it this way, getting "p" for free.